README.md

```diff
@@ -58,81 +54,52 @@ Include this badge in your readme if you make a new module that implements inter
 
 A valid (read: that follows this interface) IPLD format implementation the following API.
 
-### IPLD format utils
+IPLD Format APIs are restricted to a single IPLD Node, they never access any linked IPLD Nodes.
 
-#### `util.serialize(dagNode, callback)`
 
-> serializes a dagNode of an IPLD format into its binary format
+### `serialize(IpldNode)`
 
-`callback` must have the signature `function (err, binaryBlob)`, where `err` is an Error is the function fails and `binaryBlob` is a Buffer containing the serialized version.
+> Serializes the internal representation of an IPLD Node into a binary blob.
 
-#### `util.deserialize(binaryBlob, callback)`
+`IpldNode` is a previously deserialized binary Block.
 
-> deserializes a binary blob into the instance
+Returns a Promise containing a `Buffer` with the serialized version of the given IPLD Node.
 
-`callback` must have the signature `function (err, dagNode)`, where `err` is an Error if the function fails and `dagNode` is the dagNode that got deserialized in the process.
 
-#### `util.cid(binaryBlob[, options], callback)`
+### `deserialize(binaryBlob)`
 
-> get the CID of a binary blob
+> Deserialize into internal representation.
 
-Options include:
-  - version - the CID version to be used (defaults to 1)
-  - hashAlg - the hash algorithm to be used (default to the one set by the format)
+The result is a JavaScript object, which may also be a Proxy object in case the data shouldn’t be deserialized as a whole. Its fields are the public API that can be resolved through. It’s up to the format to add convenient methods for manipulating the data. The returned object may also be a Proxy object in case the data shouldn’t be deserialized as a whole.
```

Comment:
I did wrote more about this above but I think I want add one more note here - Requirement for implementation to provide a proxy just to avoid full decoding makes a high bar, in fact I would argue that custom [`DataView`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView) class with getters is probably a far better alternative. However that still  implies 5 allocations when accessing something like `/deeply/A0/01/FC-nested/path.txt` which may seem insignificant but in some cases (e.g. representing Virtual DOM changelists in [dominion](https://github.com/gozala/dominion) library) that makes a difference between fast and unusable performance.

Reply:
It's not a requirement, _"may also be a Proxy object"_, but I'm very negative on Proxies regardless. Aside from performance concerns, I think we've learned from many previous attempts in other languages that things like Proxies should be reserved for metaprogramming and other reflection use-cases rather than general magical data structures. Pitfalls abound.
Perhaps the way forward is to remove the reference to proxies completely and leave it up to each format and allow experimentation.

Alternatively go the container route which is much nicer--I'm not sure `DataView` is the right container, maybe a bit too low-level, something that feels more like a Map might be more accessible. In fact, a whole custom Map-like that you could `.toObject()` to fully instantiate as a non-lazy object representation might provide the best of both worlds?

```js
let data = await format.deserialize(blob)
data.get('foo') // could be lazy
data.has('/deeply/A0/01/FC-nested/path.txt') // could be lazy
data.toObject() // skip laziness and recursively instantiate
// { foo: 'bar', deeply: { A0: { '01': { 'FC-nested': { 'path.txt': UInt8Array<00, 00, 00, 00> } } } } }
```

It would require specification, and even provision of a custom class, maybe just a simple extension of `Map`.

For deserialization where you know the result is cheap and you want it all you could just instantiate it in the same call:

```js
let copy = Object.extend((await format.deserialize(blob)).toObject(), { lastUpdate: new Date() })
```

```
Map {
  get (key/path)
  has (key/path)
  toObject()
  put() // throws? is mutability helpful here? maybe until it starts to hit awkward edges
}
```

Are there possible use-cases where getters might need to be async or is it safe to frontload that in `decode()`?

Reply:
I think you may have misunderstood what I meant by `DataView` - just a container tgat wraps underlying buffer and provides getters to access / decode fields on demand `get key() {`. So pretty much what you described as Map except named getters on prototype instead of `get(key)`

Reply:
> Perhaps the way forward is to remove the reference to proxies completely and leave it up to each format and allow experimentation.
 
The whole point of IPLD is to provide common interface for traversing linked data no ? If each implementation is free to design one, the common interface is gone & I think that’s a step in a wrong direction - I much rather have common imperfect interface that can be used across arbitrary formats than have to learn each one

Reply:
> just a container tgat wraps underlying buffer and provides getters to access / decode fields on demand

But this then won't be lazy for nested objects, will it?

Reply:
It will be if for nested records you return corresponding  `DataView`

Reply:
It will be if for nested records you return corresponding  `DataView`

Reply:
And that would be faster then a JS Proxy (I haven't really used Proxies so far, I only toyed around with them)?

Reply:
Few more thoughts on this:

- `DataView` only works in cases where shape of data is static and known ahead of time like protocol buffers however it’s not for generic structures like dictionaries or JSON.
- Proxies also imply syncronous decoding otherwise you’ll have to return promise that may resolve to void or fail (makes me even more concerned)
- If you just return proxy / DataView what the point ? It seems like that can be automated by tge library and provide consistent API + reduce surface for error.

  maybe `dag.view(blob, format)` ahould just do that and utilize lower level API to materialize fields on access.

- It might be worth having two separate endpoints one to decode blob fully & the other to do that per fields. That would allow a user choose right API per use case.

Reply:
A general note on this discussion thread: My intention of these API changes is to make it easier to use. The current api has `resolve()` which is similar to what @Gozala and @rvagg are proposing. If the things that `deserialize()` returns would just be a JavaScript object, you wouldn't need to learn any new APIs, you could just use it as you're used to do and access all properties directly.

But perhaps  is treating IPLD Nodes as objects is just another layer on-top of lower level building blocks.

Reply:
I had a brief conversation with @mikeal that touched on this and it's made me wonder if we're dabbling a bit too much in over-engineering without a use-case that warrants it? I don't know enough yet about the formats so I'll defer to you both, but are there any formats that could justify a lazy parsing layer? We've been messing around with optimizing JSON parsing for years now and we've still not shown that lazy parsing provides any benefits, except perhaps in very extreme cases where the data shape is simple (perhaps predictable) and the data size is large.

So my questions are:

1. Do we imagine large IPLD blocks in common usage, or are we settling on a rough maximum of 1 or 2 Mb (that seems to be what I'm seeing, but only for filesystem type data, everything else I've seen is considerably smaller)
2. Are there any formats that are in common use that could take advantage of lazy parsing with any meaningful performance gain?
3. Are there any potential formats that could take advantage of lazy parsing for meaningful performance gain, even with Flatbuffers @Gozala, is the benefit even meaningful under 1 or 2 Mb since you still have some up-front overhead, i.e. lazy is never free?

When I talk about "meaningful" gain, it has to be in comparison to all of the other weak spots, already I'm finding that the local-only disk storage is painfully slow for simple blocks, p2p storage has orders of magnitude more performance problems, I can't imagine that the performance optimizations we're talking about can even touch the gains we're talking about getting out of this in exchange for some combination of implementation complexity or API complexity.

I notice that a Q4 target @ https://github.com/ipld/roadmap/pull/6/files#diff-04c6e90faac2675aa89e2176d2eec7d8R32 is about returning to the interfaces after "practical" use of them. So maybe it's best to keep it simple and figure out where the weak performance spots are and returning later?

Reply:
>  I don't know enough yet about the formats so I'll defer to you both, but are there any formats that could justify a lazy parsing layer?

As I mentioned previously https://github.com/google/flatbuffers/ is a great example of such data format. It's main advantage over protocol buffers is exactly due to fact that it does not require allocations to decode relevant fields. Here is also good article that goes in more details about this https://code.fb.com/android/improving-facebook-s-performance-on-android-with-flatbuffers/ and perf comparison with other formats https://google.github.io/flatbuffers/flatbuffers_benchmarks.html

Reply:
Point that I was mostly trying to make that switching to this new API comes with trade-off, specifically it prevents lazy decoding and requires allocations where former API did not had to. This can be avoided without new fancier API, and that fancier API can be provided in addition without to an API that makes no assumptions about laziness & probably would make format implementer's life easier.

Whether that is worthy trade-off is maintainers call, I'm however inclined to disagree

Reply:
@Gozala so specifically your recommendation on that is:

> It might be worth having two separate endpoints one to decode blob fully & the other to do that per fields. That would allow a user choose right API per use case.

Are you also saying that a DataView-style approach will be helpful in addition or as part of the per-field decode API?

My suggestion would be to optimise for, in rank order: (1) usability for consumer (simple and familiar feeling API), (2) usability for implementer (lower complexity) and (3) performance. At the moment I'm seeing arguments about performance to be mostly hypothetical, which is one of the main causes of over-engineering.

Reply:
I’d recommend avoiding the complexity of accounting for Proxy objects in this release. Once we have a working prototype that needs this feature we can update the interface.

This will not be the last iteration of this API :)

Reply:
> Are you also saying that a DataView-style approach will be helpful in addition or as part of the per-field decode API?

I don’t have formed opinion on this. All I was trying to piont out that JS object interface could be added as a layer on top of more low-level API that doesn’t imply debated tradeoffs

> My suggestion would be to optimise for, in rank order: (1) usability for consumer (simple and familiar feeling API)

Ok so is following more usable 

```js
let a = await deserialize(blob).nested
let b = await a.b
let c = await b.c
```

Or

```js
let c = await resolve(blob, "/nested/b/c”)
```

I’d argue that later

Reply:
> (2) usability for implementer (lower complexity)

For implementing eager case this will make no difference. For lazy case avoiding proxies and lazy getters would be significantly simpler than needing to have them

Reply:
```
let a = await deserialize(blob).nested
let b = await a.b
let c = await b.c
```

When I wrote my initial response in this thread I was trying to imagine a use-case where the properties would need to be asynchronously accessed and I couldn't see an obvious one _except_ for very large blobs that need to be stored on disk--but those wouldn't be passed in to a `deserialize(blob)` call in the first place. Maybe there's a case if there was some _really_ heavy crypto applied to sections of a blob, but even then it's questionable. Async computational tasks are usually made pointless by the boundary crossing problems and is why async methods in Node's `'crypto'` module have been mostly resisted, computationally intensive thread offloading can't be demonstrated except for very heavy work. The browser is a slightly different case because you're not maximising core use as with a server and responsiveness is important, maybe that's a worthy argument? I'm doubtful.

So again, it looks like premature optimisation in search of a use-case.

> > (2) usability for implementer (lower complexity)
>
> For implementing eager case this will make no difference. For lazy case avoiding proxies and lazy getters would be significantly simpler than needing to have them

Agreed, and it'll help make quicker progress on this PR!

Reply:
I care about JS implementation in browser context (on server go is likely a better choice anyway). In [browser crypto APIs](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto) are async (for good reasons) ignoring that seems strange to me.

And honestly it’s not about optimizations, just about not making currently well supported use cases second class.

P.S. I am working on crypto heavy use case for browsers right now and this API channge will require me to switch to handrolled solution.

Reply:
@Gozala ok this is good context, excuse my ignorance. In your use case, is it necessary that the crypto be handled at the format layer or is it reasonable to push it up to ipld itself? I was imagining https://github.com/ipld/ipld/issues/64 being something in ipld and the blocks just providing the encrypted binary data. Perhaps an encrypted block coming from a format gets transposed into an unencrypted ipld block where the key is provided but this happens at the ipld layer, not in the format. Or are you seeing a different mechanism? Again, I'm fairly new to thinking about this stuff so additional insight is appreciated!

Reply:
I've created https://github.com/ipld/interface-ipld-format/pull/51 which hopefully accommodates all the points brought up here. It's a smaller change to the current API. Let's have a discussion over there.